### PR TITLE
Implement clSetProgramSpecializationConstant

### PR DESCRIPTION
**Pull request by aarongreig**
This contribution is being made by Codeplay on behalf of Samsung.

Depends on https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/1758 (and a submodule bump which I will add here once that gets merged)

### CODE REVIEW

**Comment by aarongreig** [comment]
My llvm-spirv patch merged but only after some breaking changes that'll necessitate a clspv llvm bump, I'm not going to have time to do the testing to make that change that so I'll wait until it happens and update here when it does.

**Comment by aarongreig** [comment]
I think the remaining fail might be an issue with the CI, it looks like it's failing before the build gets to anything touched by my changes.

File: src/program.hpp [hunk 1]
```diff
@@ -142,6 +142,35 @@ enum class spec_constant
...
+    union {
+        uint8_t i8;
+        uint16_t i16;
+        uint32_t i32;
+        uint64_t i64;
+    } data;
+
+    user_spec_constant_data(std::string type, uint32_t size)
+        : type(type), size(size), set(false) {}
+    user_spec_constant_data() : type(""), size(0), set(false) {}
```
Comment (kpet) [inline src/program.hpp:158]: Why is that constructor needed? Can we remove it?

File: src/program.hpp [hunk 2]
```diff
@@ -142,6 +142,35 @@ enum class spec_constant
...
+    void init_data(size_t size, const void* value) {
+        if (type == "i8" || type == "i1") {
+            std::memcpy(&data.i8, value, size);
+        } else if (type == "i16" || type == "f16") {
+            std::memcpy(&data.i16, value, size);
+        } else if (type == "i32" || type == "f32") {
+            std::memcpy(&data.i32, value, size);
+        } else if (type == "i64" || type == "f64") {
+            std::memcpy(&data.i64, value, size);
+        }
```
Comment (kpet) [inline src/program.hpp:169]: We should probably have an else case to catch errors here or at least an assert.

File: src/api.cpp [hunk 3]
```diff
@@ -5476,13 +5483,20 @@ cl_program CLVK_API_CALL clCreateProgramWithIL(cl_context context,
...
-                 program, spec_id, spec_size, spec_value);
-    return CL_INVALID_OPERATION;
+                 prog, spec_id, spec_size, spec_value);
+
+    auto program = icd_downcast(prog);
+
+    if (!is_valid_program(program)) {
+        return CL_INVALID_PROGRAM;
+    }
+
```
Comment (kpet) [inline src/api.cpp:5498]: A few more errors need to be handled here. Please handle them or add TODOs:

• CL_INVALID_OPERATION if no devices associated with program support intermediate language programs.
• CL_COMPILER_NOT_AVAILABLE if program is created with clCreateProgramWithIL and a compiler is not available, i.e. CL_DEVICE_COMPILER_AVAILABLE specified in the Device Queries table is set to CL_FALSE.
• CL_INVALID_VALUE if spec_value is NULL.

File: src/program.cpp [hunk 4]
```diff
@@ -894,9 +894,92 @@ std::string cvk_program::prepare_build_options(const cvk_device* device) const {
...
+        delim = line.find(",");
+        auto size_string = line.substr(0, delim);
+        line.erase(0, delim + 1);
+        size_string = size_string.substr(size_string.find("=") + 1);
+        auto size = static_cast<uint32_t>(atoi(size_string.c_str()));
+
+        // We need to consume the "=" and the " " here since we aren't
+        // converting the value from a string.
+        auto type_string = line.substr(line.find("=") + 2);
+        m_user_spec_constants[id] = user_spec_constant_data(type_string, size);
```
Comment (kpet) [inline src/program.cpp:949]: The lack of error handling is making me a bit nervous. Hopefully the translator does not change its output format too frequently and never produces output that would be invalid in a way that would trip this code over. Maybe a regex would be better? I think I'm fine with no change here.

**Review by kpet — CHANGES_REQUESTED** [review]
Thanks! It's almost there, a few minor changes please (also merge conflicts need to be resolved).

**Comment by kpet** [comment]
Also, I can confirm that the macOS CI error is spurious, I'll deal with this.

**Comment by kpet** [comment]
@aarongreig Do you want to update the PR or should I take it over?

**Comment by kpet** [comment]
@aarongreig It's gone in as https://github.com/kpet/clvk/pull/503. I think I've managed to retain your authorship. Thanks for the contribution!

**Comment by aarongreig** [comment]
thanks!

**Pull request closed without merging**